source/simulator/core.cc

```diff
@@ -2077,21 +2079,25 @@ namespace aspect
           double initial_stokes_residual      = 0;
           std::vector<double> initial_composition_residual (parameters.n_compositional_fields,0);
 
-          unsigned int iteration = 0;
-
           do
             {
               assemble_advection_system(AdvectionField::temperature());
 
-              if (iteration == 0)
+              if (nonlinear_iteration == 0)
                 initial_temperature_residual = system_rhs.block(introspection.block_indices.temperature).l2_norm();
 
               const double temperature_residual = solve_advection(AdvectionField::temperature());
+              const double relative_temperature_residual = (initial_temperature_residual > 0)
+                                                           ?
+                                                           temperature_residual/initial_temperature_residual
+                                                           :
+                                                           0.0;
 
               current_linearization_point.block(introspection.block_indices.temperature)
                 = solution.block(introspection.block_indices.temperature);
               rebuild_stokes_matrix = true;
               std::vector<double> composition_residual (parameters.n_compositional_fields,0);
```

Comment:
It looks like you only need the vector of relative residuals below, not the absolute residuals. So you could remove this vector and change the lines below as indicated.